pkg/reconciler/v1alpha1/route/reconcile_resources.go

```diff
@@ -131,3 +137,59 @@ func (c *Reconciler) updateStatus(ctx context.Context, route *v1alpha1.Route) (*
 	c.Recorder.Eventf(route, corev1.EventTypeNormal, "Updated", "Updated status for route %q", route.Name)
 	return updated, nil
 }
+
+// Update the lastPinned annotation on revisions we target so they don't get GC'd.
+func (c *Reconciler) reconcileTargetRevisions(ctx context.Context, t *traffic.TrafficConfig, route *v1alpha1.Route) error {
+	rev_errs := []error{}
+	for _, target := range t.Targets {
+		for _, rt := range target {
+			tt := rt.TrafficTarget
+			rev, err := c.revisionLister.Revisions(route.Namespace).Get(tt.RevisionName)
+			if apierrs.IsNotFound(err) {
+				c.Logger.Infof("Unable to update lastPinned for missing revision %q", tt.RevisionName)
+				continue
+			} else if err != nil {
+				rev_errs = append(rev_errs, err)
+				continue
+			}
+
+			// Enforce a delay before performing an update on lastPinned to avoid excess churn
+			if rev.Annotations == nil {
+				rev.Annotations = make(map[string]string)
+			}
+
+			lastPin, err := rev.GetLastPinned()
+			if err != nil && err.(v1alpha1.LastPinnedParseError).Type == v1alpha1.AnnotationParseErrorTypeInvalid {
```

Comment:
I don't like this type assertion of exclusion.  Why would we do this?

Reply:
I dont understand "type assertion of exclustion". The typecast to `v1alpha1.LastPinnedParseError` is to get the .type property; the check is to determine if there was an unparsable string for lastPinned.

Reply:
I think I meant to say "or exclusion" meaning: the exclusion of certain error types from propagating to the caller